File: 199_binary_tree_right_side_view/Solution.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {};
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public static TreeNode createTreeUsingBFS(int[] nums) {
        TreeNode[] nodeArr = new TreeNode[nums.length];
        for (int i = 0; i < nums.length; i++) {
            nodeArr[i] = new TreeNode(nums[i]);
            if (i == 0) continue;
            if (i % 2 == 1) nodeArr[(i - 1) / 2].left = nodeArr[i];
            else nodeArr[(i - 1) / 2].right = nodeArr[i];
        }

        return nodeArr[0];
    }

    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) return result;

        Deque<TreeNode> queue = new ArrayDeque<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            for (int i = queue.size(); i > 0; i--) {
                TreeNode curNode = queue.pollFirst();
                if (i == 1) result.add(curNode.val);
                if (curNode.left != null) queue.add(curNode.left);
                if (curNode.right != null) queue.add(curNode.right);
            }
        }

        return result;
    }

    public static void main(String[] args) {
        TreeNode root = createTreeUsingBFS(new int[]{10, 5, 15, 2, 6, 11, 16, 1, 3});
        System.out.println(new Solution().rightSideView(root));
    }
}

File: 199_binary_tree_right_side_view/Solution.go
package main

import "fmt"

type TreeNode struct {
    Val   int
    Left  *TreeNode
    Right *TreeNode
}

func createTreeUsingBFS(nums []int) *TreeNode {
    arrLen := len(nums)
    var nodeArr []*TreeNode

    for i := 0; i < arrLen; i++ {
        nodeArr = append(nodeArr, &TreeNode{nums[i], nil, nil})
        if i == 0 { continue }
        if i % 2 == 1 {
            nodeArr[(i - 1) / 2].Left = nodeArr[i]
        } else {
            nodeArr[(i - 1) / 2].Right = nodeArr[i]
        }
    }

    return nodeArr[0]
}

func rightSideView(root *TreeNode) []int {
    if root == nil { return []int{} }

    var result []int
    queue := []*TreeNode{root}
    for len(queue) != 0 {
        for i := len(queue); i > 0; i-- {
            curNode := queue[0]
            queue = queue[1:]
            if i == 1 { result = append(result, curNode.Val) }
            if curNode.Left != nil { queue = append(queue, curNode.Left) }
            if curNode.Right != nil { queue = append(queue, curNode.Right) }
        }
    }

    return result
}

func main() {
    nums := []int{10, 5, 15, 2, 6, 11, 16, 1, 3}
    root := createTreeUsingBFS(nums)
    fmt.Println(rightSideView(root))
}

